### ISSUE TITLE
Ingress Error: New Install of Hass.io and Node-Red - 502: Bad Gateway

### ISSUE THREAD

**Issue by danner26** [hassio-addons/addon-node-red#260]
# Problem/Motivation
Hello, so I am having a very frustrating issue. Node-red is installed properly via the hass.io add-on. I configured the software as follows:
![image](https://user-images.githubusercontent.com/8165027/60526713-d99a5400-9cbe-11e9-9bf2-5cd90b4729c7.png)

So I started troubleshooting where I thought best, I checked to make sure the docker container was running. It is running and the logs have no errors. Here is the pastebin link for the node-red logs: https://pastebin.com/54RwPGLh

So next I wanted to make sure that Node-RED was actually running, even though it says it is in the logs. So I open port 1880 on my server, and head over to 192.168.1.x:1880 and it lets me in. The UI is loaded fine.

So next I started reading around and found a few articles. The first was about an old local storage cache, well that is not my issue because this is a brand new install, and I have never used node-red on this computer. Just for the sake of it, I double checked and I was correct no storage items, so that was a dead end.

Next I was thinking and decided maybe the HA/supervisor logs had something for me, and sure enough in the supervisor logs I found the following line occured every time I navigated to node-red from the hassio interface:
```19-07-02 15:51:19 ERROR (MainThread) [hassio.api.ingress] Ingress error: Cannot connect to host 172.30.32.1:62819 ssl:None [Host is unreachable]```

Here is a copy of the supervisor log in pastebin: https://pastebin.com/jredy7zJ
A few errors I see, not sure if these are related or even matter:
```
19-07-02 15:01:33 ERROR (MainThread) [hassio.utils.gdbus] DBus return error: b'Error: Timeout was reached\n'
19-07-02 15:01:33 WARNING (MainThread) [hassio.dbus.hostname] Can't connect to hostname

19-07-02 15:01:33 ERROR (MainThread) [hassio.utils.gdbus] DBus return error: b'Error: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name de.pengutronix.rauc was not provided by any .service files\n'
19-07-02 15:01:33 WARNING (MainThread) [hassio.dbus.rauc] Can't connect to rauc
```
Here is a copy of the home assistant log in pastebin: https://pastebin.com/6ub6Ui76

Well okay, maybe something got screwy during the docker setup. Lets check the routes and make sure the interfaces are configured properly:
```
[root@TCH-HA homeassistant]# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         192.168.1.1     0.0.0.0         UG    100    0        0 ens192
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.30.32.0     0.0.0.0         255.255.254.0   U     0      0        0 hassio
192.168.1.0     0.0.0.0         255.255.255.0   U     100    0        0 ens192
```

Yup, routes are fine that cant be the issue either.. Then it struck me, I am working behind a reverse proxy that proxies the traffic over to the HA box. So let's eliminate that and just connect locally. Maybe that is the issue! I head over to http://192.168.1.x:8123/a0d7b954_nodered and unfortunately the same thing. I saw on another forum that some people get it to work using the Open Web UI button, but no luck here still a 502 error.

I am at the end of the straw and can't seem to find another possible solution. Any help would be appreciated.

## Expected behavior
Obviously we expect that the node-red interface will appear in either the sidebar or open web ui link.

## Actual behavior
502: Bad Gateway on Open Web UI or the side bar, but can connect to the node-red UI without using HA by going to 192.168.1.x:1880

**Comment by danner26** [comment]
So just a thought was that docker wasnt exposing the container. I first tried to ping 172.30.32.1 and that worked.

So next I tried telnet to the port it seemed to want to reach: 62819 and it connected fine. So the port is accessible and the host is up. Hassio locally (connecting via ip) returns a 502 error as well as the proxy'd connection.

**Comment by ernestgwilsonii** [comment]
I am having similar 502 error and on Raspberry Pi 3 using "Buster" OS, the container fails ti even start.
Could this be an AppArmor issue?

**Comment by sinclairpaul** [comment]
Just to confirm essentially the issue is the Ingress link fails, but the addon works when exposed?
I am assuming it won't be limited to Node-Red, but likely anything that runs on the host network, could you confirm by testing the Community SSH or Adguard addons (or any others that run on the host). Whats the output of ```docker network ls```

**Comment by danner26** [comment]
Yes, the issue is 100% with ingress I believe. Other addons work with ingress, if that is what you are asking. Node-RED does start and is accessible via direct connection to the web service, but like said previously ingress is not working.

Here is the output:
```
[ha@TCH-HA ~]$ docker network ls
NETWORK ID          NAME                DRIVER              SCOPE
db1ac4e32256        bridge              bridge              local
439abfe4eadf        hassio              bridge              local
e2b582284517        host                host                local
7f1e13e5bd4f        none                null                local
```

**Comment by sinclairpaul** [comment]
Please confirm using the Community SSH/Web Terminal addon the behaviour, from your log non of the other addons you have run on the host network.  This line in the logs is concerning (the null part):

``` System: null  (amd64 / qemux86-64)```

Whats your distro?, and would you happen to have any form of firewall running.  This is very unlikely to be an issue with the Node-Red addon itself, and likely something at the host level.

**Comment by danner26** [comment]
I see, okay yeah the terminal plugin is not working as well. Let me provide you some more info:
```
[root@TCH-HA ~]# cat /etc/centos-release
CentOS Linux release 7.6.1810 (Core)

[root@TCH-HA ~]# yum list installed | grep firewall
firewalld.noarch                     0.5.3-5.el7                 @base
firewalld-filesystem.noarch          0.5.3-5.el7                 @base
python-firewall.noarch               0.5.3-5.el7                 @base
```

CentOS 7 with firewalld enabled. Here are my firewall rules, but since the connection should be local for ingress, I dont see the firewall being the issue. I am more leaning towards a Docker network issue, but maybe you have more insight than I do.
```
[root@TCH-HA ~]# firewall-cmd --list-all
public (active)
  target: default
  icmp-block-inversion: no
  interfaces: ens192
  sources:
  services: dhcpv6-client ssh
  ports: 8123/tcp 8123/udp 1880/tcp 1880/udp 4865/udp 4865/tcp 53/tcp 53/udp
  protocols:
  masquerade: no
  forward-ports:
  source-ports:
  icmp-blocks:
  rich rules:
```

**Comment by frenck** [comment]
Check your logs to see if your firewall blocks traffic from/to the `hassio` Docker network.

**Comment by danner26** [comment]
You are right, okay progress. I disabled the firewall and everything started working. So now the question is, what firewall rules need to be created in order for this to work?

**Comment by danner26** [comment]
Okay, I see in the firewalld logs errors like these:
```
2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -n -L DOCKER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -n -L DOCKER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -n -L DOCKER-ISOLATION-STAGE-1' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -n -L DOCKER-ISOLATION-STAGE-2' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C DOCKER-ISOLATION-STAGE-1 -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C DOCKER-ISOLATION-STAGE-2 -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C POSTROUTING -s 172.30.32.0/23 ! -o hassio -j MASQUERADE' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C DOCKER -i hassio -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -D FORWARD -i hassio -o hassio -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -i hassio -o hassio -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -i hassio ! -o hassio -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C PREROUTING -m addrtype --dst-type LOCAL -j DOCKER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C OUTPUT -m addrtype --dst-type LOCAL -j DOCKER ! --dst 127.0.0.0/8' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -o hassio -j DOCKER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -o hassio -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -j DOCKER-ISOLATION-STAGE-1' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C DOCKER -p tcp -d 0/0 --dport 3000 -j DNAT --to-destination 172.30.33.0:80 ! -i hassio' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C DOCKER ! -i hassio -o hassio -p tcp -d 172.30.33.0 --dport 80 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C POSTROUTING -p tcp -s 172.30.33.0 -d 172.30.33.0 --dport 80 -j MASQUERADE' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C DOCKER -p tcp -d 0/0 --dport 1627 -j DNAT --to-destination 172.30.33.2:1627 ! -i hassio' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C DOCKER ! -i hassio -o hassio -p tcp -d 172.30.33.2 --dport 1627 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C POSTROUTING -p tcp -s 172.30.33.2 -d 172.30.33.2 --dport 1627 -j MASQUERADE' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C POSTROUTING -s 172.17.0.0/16 ! -o docker0 -j MASQUERADE' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t nat -C DOCKER -i docker0 -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -i docker0 -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -i docker0 ! -o docker0 -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -o docker0 -j DOCKER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:15 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -o docker0 -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:16 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -D FORWARD -i docker0 -o docker0 -j DROP' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:16 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -n -L DOCKER-USER' failed: iptables: No chain/target/match by that name.

2019-07-08 09:47:16 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C DOCKER-USER -j RETURN' failed: iptables: Bad rule (does a matching rule exist in that chain?).

2019-07-08 09:47:16 WARNING: COMMAND_FAILED: '/usr/sbin/iptables -w2 -t filter -C FORWARD -j DOCKER-USER' failed: iptables: No chain/target/match by that name.
```

I will start looking at solutions to these issues, I am guessing this is the cause?

**Comment by sinclairpaul** [comment]
Yes, and my google skills show me this may help https://success.docker.com/article/firewalld-problems-with-container-to-container-network-communications (though I doubt the port is relevant).

**Comment by danner26** [comment]
I did already try that, thank you though. I now see it is definitely a firewall issue so I can go through troubleshooting that.. Once I find a solution I will report back

**Comment by f4t0** [comment]
Same Error. Just wait 1 minute and then it works!

**Comment by frenck** [comment]
@f4t0 a blocking firewall is not going to fix itself...

**Comment by danner26** [comment]
Agreed, @f4t0 I have a different issue

**Comment by danner26** [comment]
I noticed that the more I messed with firewall rules on CentOS, the more that broke. Ingress for previously working things broke further. 

I got fed up, made a backup of the OS and installed Ubuntu Server 18 on the machine, reinstalled hass.io and docker, and everything just worked out of the box. Maybe it was an issue with looking for AppArmour (which doesnt exist in CentOS land, its selinux for us). So for now, since everything is working on Ubuntu I am going to stick with that. 

Thanks for anyone that tried to help.
